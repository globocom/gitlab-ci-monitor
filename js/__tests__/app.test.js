import axios from 'axios';
import MockAdapter from 'axios-mock-adapter';

import projectMock from '../__mocks__/project.json';
import buildsMock from '../__mocks__/builds.json';
import commitMock from '../__mocks__/commit.json';
import pipelineMock from '../__mocks__/pipeline.json';

jest.useFakeTimers();
Date.now = jest.fn(() => 1538362800000);

async function createVm() {
  return require('../app').default;
}

function setupAxiosMock() {
  const mock = new MockAdapter(axios);
  mock.onGet('/projects/namespace%2Fproject1').reply(200, projectMock);
  mock.onGet('/projects/namespace%2Fproject2').reply(200, projectMock);
  mock.onGet('/projects/6944716/pipelines/?ref=master').reply(200, buildsMock);
  mock.onGet('/projects/6944716/pipelines/?ref=branch1').reply(200, buildsMock);
  mock.onGet('/projects/6944716/pipelines/23827851').reply(200, pipelineMock);
  mock.onGet('/projects/6944716/repository/commits/25fdf4ca24b7a97e9a895abc43fc54128f218ddd').reply(200, commitMock);
}

describe('Testing app.js', () => {
  let app;

  beforeAll(async () => {
    setupAxiosMock();
    document.body.innerHTML = '<div id="app"></div>';
    app = await createVm();
  });

  test('snapshotting', () => {
    expect(app).toMatchSnapshot();
  });

  test('app.methods.loadConfig', () => {
    const {
      gitlab,
      token,
      repositories,
      sortFields
    } = app;
    expect(gitlab).toEqual('gitlab.test.com');
    expect(token).toEqual('12345');

    // Validate SortFields
    expect(sortFields.length).toEqual(1);
    expect(sortFields[0]).toEqual({ field: 'project', dir: 'asc' });

    // Validate Repositories
    expect(repositories.length).toEqual(2);
    expect(repositories[0]).toEqual({
      key: "namespace/project1/branch1",
      projectName: "project1",
      branch: 'branch1',
      nameWithNamespace: "namespace/project1",
    });
    expect(repositories[1]).toEqual({
      key: "namespace/project2/master",
      projectName: "project2",
      branch: 'master',
      nameWithNamespace: "namespace/project2",
    });
  });

  test('app.methods.fetchProjects', () => {
    expect(app.projects['namespace/project1/branch1'].data).toEqual(projectMock);
    expect(app.projects['namespace/project2/master'].data).toEqual(projectMock);
  });

  test('app.methods.updateBuilds', async () => {
    const now = app.lastRun;
    jest.spyOn(app, 'updateBuilds');
    jest.spyOn(Date, 'now').mockImplementationOnce(() => 1540436400000);
    expect(app.updateBuilds).toHaveBeenCalledTimes(0);
    jest.advanceTimersByTime(70000);
    expect(app.updateBuilds).toHaveBeenCalledTimes(1);
    const then = app.lastRun;
    expect(now).not.toEqual(then);
  });
});