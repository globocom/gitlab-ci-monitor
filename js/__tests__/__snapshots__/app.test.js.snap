// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing app.js snapshotting 1`] = `
Vue {
  "$children": Array [],
  "$el": <div
    id="app"
  />,
  "$els": Object {},
  "$options": Object {
    "_containerAttrs": null,
    "_replacerAttrs": null,
    "components": Object {},
    "computed": Object {
      "sortedPipelines": [Function],
    },
    "created": Array [
      [Function],
    ],
    "data": [Function],
    "directives": Object {},
    "el": <div
      id="app"
    />,
    "elementDirectives": Object {},
    "filters": Object {},
    "methods": Object {
      "fetchBuild": [Function],
      "fetchGroups": [Function],
      "fetchProjects": [Function],
      "loadConfig": [Function],
      "setupDefaults": [Function],
      "updateBuildInfo": [Function],
      "updateBuilds": [Function],
      "validateConfig": [Function],
    },
    "partials": Object {},
    "replace": true,
    "transitions": Object {},
  },
  "$parent": undefined,
  "$refs": Object {},
  "$root": [Circular],
  "_context": undefined,
  "_data": Object {
    "gitlab": "gitlab.test.com",
    "invalidConfig": false,
    "lastRun": "2018-10-01 - 00:00:00",
    "loading": false,
    "onError": null,
    "orderFields": Object {
      "dir": "asc",
      "field": "project",
    },
    "pipelines": Array [],
    "pipelinesMap": Object {},
    "projects": Object {
      "namespace/project1/branch1": Object {
        "data": Object {
          "_links": Object {
            "events": "https://gitlab.com/api/v4/projects/6944716/events",
            "issues": "https://gitlab.com/api/v4/projects/6944716/issues",
            "labels": "https://gitlab.com/api/v4/projects/6944716/labels",
            "members": "https://gitlab.com/api/v4/projects/6944716/members",
            "merge_requests": "https://gitlab.com/api/v4/projects/6944716/merge_requests",
            "repo_branches": "https://gitlab.com/api/v4/projects/6944716/repository/branches",
            "self": "https://gitlab.com/api/v4/projects/6944716",
          },
          "approvals_before_merge": 0,
          "archived": false,
          "avatar_url": "https://assets.gitlab-static.net/uploads/-/system/project/avatar/6944716/81NBTg9OS3L._SL1500_.jpg",
          "ci_config_path": null,
          "container_registry_enabled": true,
          "created_at": "2018-06-07T23:42:54.228Z",
          "creator_id": 745415,
          "default_branch": "master",
          "description": "Projeto de exemplo para demonstrar integração de ferramentas em um projeto JS.",
          "forks_count": 0,
          "http_url_to_repo": "https://gitlab.com/gvarandas/bsb-dev-festival-ci.git",
          "id": 6944716,
          "import_error": null,
          "import_status": "none",
          "issues_enabled": true,
          "jobs_enabled": true,
          "last_activity_at": "2018-06-14T04:26:18.770Z",
          "lfs_enabled": false,
          "merge_method": "merge",
          "merge_requests_enabled": true,
          "mirror": false,
          "name": "bsb-dev-festival-ci",
          "name_with_namespace": "Guilherme Varandas / bsb-dev-festival-ci",
          "namespace": Object {
            "full_path": "gvarandas",
            "id": 891221,
            "kind": "user",
            "name": "gvarandas",
            "parent_id": null,
            "path": "gvarandas",
          },
          "only_allow_merge_if_all_discussions_are_resolved": false,
          "only_allow_merge_if_pipeline_succeeds": false,
          "open_issues_count": 0,
          "owner": Object {
            "avatar_url": "https://assets.gitlab-static.net/uploads/-/system/user/avatar/745415/avatar.png",
            "id": 745415,
            "name": "Guilherme Varandas",
            "state": "active",
            "username": "gvarandas",
            "web_url": "https://gitlab.com/gvarandas",
          },
          "path": "bsb-dev-festival-ci",
          "path_with_namespace": "gvarandas/bsb-dev-festival-ci",
          "permissions": Object {
            "group_access": null,
            "project_access": Object {
              "access_level": 40,
              "notification_level": 3,
            },
          },
          "printing_merge_request_link_enabled": true,
          "public_jobs": true,
          "readme_url": "https://gitlab.com/gvarandas/bsb-dev-festival-ci/blob/master/README.md",
          "request_access_enabled": false,
          "resolve_outdated_diff_discussions": false,
          "runners_token": "SoaFQPsLMbkybHUX-XTQ",
          "shared_runners_enabled": true,
          "shared_with_groups": Array [],
          "snippets_enabled": true,
          "ssh_url_to_repo": "git@gitlab.com:gvarandas/bsb-dev-festival-ci.git",
          "star_count": 3,
          "tag_list": Array [],
          "visibility": "public",
          "web_url": "https://gitlab.com/gvarandas/bsb-dev-festival-ci",
          "wiki_enabled": false,
        },
        "project": Object {
          "branch": "branch1",
          "key": "namespace/project1/branch1",
          "nameWithNamespace": "namespace/project1",
          "projectName": "project1",
        },
      },
      "namespace/project2/master": Object {
        "data": Object {
          "_links": Object {
            "events": "https://gitlab.com/api/v4/projects/6944716/events",
            "issues": "https://gitlab.com/api/v4/projects/6944716/issues",
            "labels": "https://gitlab.com/api/v4/projects/6944716/labels",
            "members": "https://gitlab.com/api/v4/projects/6944716/members",
            "merge_requests": "https://gitlab.com/api/v4/projects/6944716/merge_requests",
            "repo_branches": "https://gitlab.com/api/v4/projects/6944716/repository/branches",
            "self": "https://gitlab.com/api/v4/projects/6944716",
          },
          "approvals_before_merge": 0,
          "archived": false,
          "avatar_url": "https://assets.gitlab-static.net/uploads/-/system/project/avatar/6944716/81NBTg9OS3L._SL1500_.jpg",
          "ci_config_path": null,
          "container_registry_enabled": true,
          "created_at": "2018-06-07T23:42:54.228Z",
          "creator_id": 745415,
          "default_branch": "master",
          "description": "Projeto de exemplo para demonstrar integração de ferramentas em um projeto JS.",
          "forks_count": 0,
          "http_url_to_repo": "https://gitlab.com/gvarandas/bsb-dev-festival-ci.git",
          "id": 6944716,
          "import_error": null,
          "import_status": "none",
          "issues_enabled": true,
          "jobs_enabled": true,
          "last_activity_at": "2018-06-14T04:26:18.770Z",
          "lfs_enabled": false,
          "merge_method": "merge",
          "merge_requests_enabled": true,
          "mirror": false,
          "name": "bsb-dev-festival-ci",
          "name_with_namespace": "Guilherme Varandas / bsb-dev-festival-ci",
          "namespace": Object {
            "full_path": "gvarandas",
            "id": 891221,
            "kind": "user",
            "name": "gvarandas",
            "parent_id": null,
            "path": "gvarandas",
          },
          "only_allow_merge_if_all_discussions_are_resolved": false,
          "only_allow_merge_if_pipeline_succeeds": false,
          "open_issues_count": 0,
          "owner": Object {
            "avatar_url": "https://assets.gitlab-static.net/uploads/-/system/user/avatar/745415/avatar.png",
            "id": 745415,
            "name": "Guilherme Varandas",
            "state": "active",
            "username": "gvarandas",
            "web_url": "https://gitlab.com/gvarandas",
          },
          "path": "bsb-dev-festival-ci",
          "path_with_namespace": "gvarandas/bsb-dev-festival-ci",
          "permissions": Object {
            "group_access": null,
            "project_access": Object {
              "access_level": 40,
              "notification_level": 3,
            },
          },
          "printing_merge_request_link_enabled": true,
          "public_jobs": true,
          "readme_url": "https://gitlab.com/gvarandas/bsb-dev-festival-ci/blob/master/README.md",
          "request_access_enabled": false,
          "resolve_outdated_diff_discussions": false,
          "runners_token": "SoaFQPsLMbkybHUX-XTQ",
          "shared_runners_enabled": true,
          "shared_with_groups": Array [],
          "snippets_enabled": true,
          "ssh_url_to_repo": "git@gitlab.com:gvarandas/bsb-dev-festival-ci.git",
          "star_count": 3,
          "tag_list": Array [],
          "visibility": "public",
          "web_url": "https://gitlab.com/gvarandas/bsb-dev-festival-ci",
          "wiki_enabled": false,
        },
        "project": Object {
          "branch": "master",
          "key": "namespace/project2/master",
          "nameWithNamespace": "namespace/project2",
          "projectName": "project2",
        },
      },
    },
    "repositories": Array [
      Object {
        "branch": "branch1",
        "key": "namespace/project1/branch1",
        "nameWithNamespace": "namespace/project1",
        "projectName": "project1",
      },
      Object {
        "branch": "master",
        "key": "namespace/project2/master",
        "nameWithNamespace": "namespace/project2",
        "projectName": "project2",
      },
    ],
    "token": "12345",
  },
  "_directives": Array [],
  "_events": Object {
    "hook:attached": Array [
      [Function],
      [Function],
    ],
    "hook:detached": Array [
      [Function],
    ],
  },
  "_eventsCount": Object {},
  "_frag": undefined,
  "_fragment": null,
  "_fragmentEnd": null,
  "_fragmentStart": null,
  "_isAttached": false,
  "_isBeingDestroyed": false,
  "_isCompiled": true,
  "_isDestroyed": false,
  "_isFragment": false,
  "_isReady": false,
  "_isVue": true,
  "_propsUnlinkFn": null,
  "_scope": undefined,
  "_uid": 0,
  "_unlinkFn": [Function],
  "_vForRemoving": false,
  "_watchers": Array [
    Watcher {
      "active": true,
      "cb": null,
      "depIds": Set {},
      "deps": Array [],
      "dirty": true,
      "expression": [Function],
      "getter": [Function],
      "id": 1,
      "lazy": true,
      "newDepIds": Set {},
      "newDeps": Array [],
      "prevError": null,
      "queued": false,
      "setter": undefined,
      "shallow": false,
      "value": undefined,
      "vm": [Circular],
    },
  ],
  "fetchBuild": [Function],
  "fetchGroups": [Function],
  "fetchProjects": [Function],
  "gitlab": "gitlab.test.com",
  "groups": Array [],
  "invalidConfig": false,
  "lastRun": "2018-10-01 - 00:00:00",
  "loadConfig": [Function],
  "loading": false,
  "onError": null,
  "orderFields": Object {
    "dir": "asc",
    "field": "project",
  },
  "pipelines": Array [],
  "pipelinesMap": Object {},
  "projects": Object {
    "namespace/project1/branch1": Object {
      "data": Object {
        "_links": Object {
          "events": "https://gitlab.com/api/v4/projects/6944716/events",
          "issues": "https://gitlab.com/api/v4/projects/6944716/issues",
          "labels": "https://gitlab.com/api/v4/projects/6944716/labels",
          "members": "https://gitlab.com/api/v4/projects/6944716/members",
          "merge_requests": "https://gitlab.com/api/v4/projects/6944716/merge_requests",
          "repo_branches": "https://gitlab.com/api/v4/projects/6944716/repository/branches",
          "self": "https://gitlab.com/api/v4/projects/6944716",
        },
        "approvals_before_merge": 0,
        "archived": false,
        "avatar_url": "https://assets.gitlab-static.net/uploads/-/system/project/avatar/6944716/81NBTg9OS3L._SL1500_.jpg",
        "ci_config_path": null,
        "container_registry_enabled": true,
        "created_at": "2018-06-07T23:42:54.228Z",
        "creator_id": 745415,
        "default_branch": "master",
        "description": "Projeto de exemplo para demonstrar integração de ferramentas em um projeto JS.",
        "forks_count": 0,
        "http_url_to_repo": "https://gitlab.com/gvarandas/bsb-dev-festival-ci.git",
        "id": 6944716,
        "import_error": null,
        "import_status": "none",
        "issues_enabled": true,
        "jobs_enabled": true,
        "last_activity_at": "2018-06-14T04:26:18.770Z",
        "lfs_enabled": false,
        "merge_method": "merge",
        "merge_requests_enabled": true,
        "mirror": false,
        "name": "bsb-dev-festival-ci",
        "name_with_namespace": "Guilherme Varandas / bsb-dev-festival-ci",
        "namespace": Object {
          "full_path": "gvarandas",
          "id": 891221,
          "kind": "user",
          "name": "gvarandas",
          "parent_id": null,
          "path": "gvarandas",
        },
        "only_allow_merge_if_all_discussions_are_resolved": false,
        "only_allow_merge_if_pipeline_succeeds": false,
        "open_issues_count": 0,
        "owner": Object {
          "avatar_url": "https://assets.gitlab-static.net/uploads/-/system/user/avatar/745415/avatar.png",
          "id": 745415,
          "name": "Guilherme Varandas",
          "state": "active",
          "username": "gvarandas",
          "web_url": "https://gitlab.com/gvarandas",
        },
        "path": "bsb-dev-festival-ci",
        "path_with_namespace": "gvarandas/bsb-dev-festival-ci",
        "permissions": Object {
          "group_access": null,
          "project_access": Object {
            "access_level": 40,
            "notification_level": 3,
          },
        },
        "printing_merge_request_link_enabled": true,
        "public_jobs": true,
        "readme_url": "https://gitlab.com/gvarandas/bsb-dev-festival-ci/blob/master/README.md",
        "request_access_enabled": false,
        "resolve_outdated_diff_discussions": false,
        "runners_token": "SoaFQPsLMbkybHUX-XTQ",
        "shared_runners_enabled": true,
        "shared_with_groups": Array [],
        "snippets_enabled": true,
        "ssh_url_to_repo": "git@gitlab.com:gvarandas/bsb-dev-festival-ci.git",
        "star_count": 3,
        "tag_list": Array [],
        "visibility": "public",
        "web_url": "https://gitlab.com/gvarandas/bsb-dev-festival-ci",
        "wiki_enabled": false,
      },
      "project": Object {
        "branch": "branch1",
        "key": "namespace/project1/branch1",
        "nameWithNamespace": "namespace/project1",
        "projectName": "project1",
      },
    },
    "namespace/project2/master": Object {
      "data": Object {
        "_links": Object {
          "events": "https://gitlab.com/api/v4/projects/6944716/events",
          "issues": "https://gitlab.com/api/v4/projects/6944716/issues",
          "labels": "https://gitlab.com/api/v4/projects/6944716/labels",
          "members": "https://gitlab.com/api/v4/projects/6944716/members",
          "merge_requests": "https://gitlab.com/api/v4/projects/6944716/merge_requests",
          "repo_branches": "https://gitlab.com/api/v4/projects/6944716/repository/branches",
          "self": "https://gitlab.com/api/v4/projects/6944716",
        },
        "approvals_before_merge": 0,
        "archived": false,
        "avatar_url": "https://assets.gitlab-static.net/uploads/-/system/project/avatar/6944716/81NBTg9OS3L._SL1500_.jpg",
        "ci_config_path": null,
        "container_registry_enabled": true,
        "created_at": "2018-06-07T23:42:54.228Z",
        "creator_id": 745415,
        "default_branch": "master",
        "description": "Projeto de exemplo para demonstrar integração de ferramentas em um projeto JS.",
        "forks_count": 0,
        "http_url_to_repo": "https://gitlab.com/gvarandas/bsb-dev-festival-ci.git",
        "id": 6944716,
        "import_error": null,
        "import_status": "none",
        "issues_enabled": true,
        "jobs_enabled": true,
        "last_activity_at": "2018-06-14T04:26:18.770Z",
        "lfs_enabled": false,
        "merge_method": "merge",
        "merge_requests_enabled": true,
        "mirror": false,
        "name": "bsb-dev-festival-ci",
        "name_with_namespace": "Guilherme Varandas / bsb-dev-festival-ci",
        "namespace": Object {
          "full_path": "gvarandas",
          "id": 891221,
          "kind": "user",
          "name": "gvarandas",
          "parent_id": null,
          "path": "gvarandas",
        },
        "only_allow_merge_if_all_discussions_are_resolved": false,
        "only_allow_merge_if_pipeline_succeeds": false,
        "open_issues_count": 0,
        "owner": Object {
          "avatar_url": "https://assets.gitlab-static.net/uploads/-/system/user/avatar/745415/avatar.png",
          "id": 745415,
          "name": "Guilherme Varandas",
          "state": "active",
          "username": "gvarandas",
          "web_url": "https://gitlab.com/gvarandas",
        },
        "path": "bsb-dev-festival-ci",
        "path_with_namespace": "gvarandas/bsb-dev-festival-ci",
        "permissions": Object {
          "group_access": null,
          "project_access": Object {
            "access_level": 40,
            "notification_level": 3,
          },
        },
        "printing_merge_request_link_enabled": true,
        "public_jobs": true,
        "readme_url": "https://gitlab.com/gvarandas/bsb-dev-festival-ci/blob/master/README.md",
        "request_access_enabled": false,
        "resolve_outdated_diff_discussions": false,
        "runners_token": "SoaFQPsLMbkybHUX-XTQ",
        "shared_runners_enabled": true,
        "shared_with_groups": Array [],
        "snippets_enabled": true,
        "ssh_url_to_repo": "git@gitlab.com:gvarandas/bsb-dev-festival-ci.git",
        "star_count": 3,
        "tag_list": Array [],
        "visibility": "public",
        "web_url": "https://gitlab.com/gvarandas/bsb-dev-festival-ci",
        "wiki_enabled": false,
      },
      "project": Object {
        "branch": "master",
        "key": "namespace/project2/master",
        "nameWithNamespace": "namespace/project2",
        "projectName": "project2",
      },
    },
  },
  "ref": null,
  "repositories": Array [
    Object {
      "branch": "branch1",
      "key": "namespace/project1/branch1",
      "nameWithNamespace": "namespace/project1",
      "projectName": "project1",
    },
    Object {
      "branch": "master",
      "key": "namespace/project2/master",
      "nameWithNamespace": "namespace/project2",
      "projectName": "project2",
    },
  ],
  "setupDefaults": [Function],
  "sortFields": Array [
    Object {
      "dir": "asc",
      "field": "project",
    },
  ],
  "sortedPipelines": Array [],
  "token": "12345",
  "updateBuildInfo": [Function],
  "updateBuilds": [Function],
  "validateConfig": [Function],
}
`;
